### ISSUE TITLE
Spacemacs hangs (CPU sticks at 100%) when doing :ex Search and Replace with global

### ISSUE THREAD

**Issue by scrotty** [syl20bnr/spacemacs#8534]
#### Description :octocat:
Spacemacs hangs (CPU sticks at 100%) when doing :ex Search and Replace with global

#### Reproduction guide :beetle:
- Start Emacs
- In scratch buffer place the following line: A|B|C|D|E|F|G|H|I|J|K|L|M|N|O|P|Q|R|S|T|U|V|W|X|Y|Z|A|B|C|D|E|F|G|H|I|J|K|L|M|N|O|P|Q|R|S|T|U|V|W|X|Y|Z|A|B|C|D|E|F|G|H|I|J|K|L|M|N|O|P|Q|R|S|T|U|V|W|X|Y|Z|A|B|C|D|E|F|G|H|I|J|K|L|M|N|O|P|Q|R|S|T|U|V|W|X|Y|Z|A|B|C|D|E|F|G|H|I|J|K|L|M|N|O|P|Q|R|S|T|U|V|W|X|Y|Z|A|B|C|D|E|F|G|H|I|J|K|L|M|N|O|P|Q|R|S|T|U|V|W|X|Y|Z|A|B|C|D|E|F|G|H|I|J|K|L|M|N|O|P|Q|R|S|T|U|V|W|X|Y|Z
- Then, in the :ex command line, do `:%s/|/\n/g` (This ex command will try to replace all pipe characters with a newline)

*Observed behaviour:* :eyes: :broken_heart:
For lines with fewer pipes, things work as expected. But when the number of pipes (or, probably more accurately, the number of global substitutions per line) hits a certain threshold, emacs hangs. This happens _before_ actually executing the search-n-replace. It happens when the "here's what will happen if you execute the command" is displayed.

By the way, It doesn't seem that there is an exact number of replacements that force the problem. Sometimes a line that hangs spacemacs most of the time will occasionally work. (I know this makes the problem much harder to pin down!) So if you have trouble reproducing, try increasing the length of the line and number of pipes.

*Expected behaviour:* :heart: :smile:
The replace preview should work properly (doesn't hang the system) and the actual execution of the replace will work properly too.

#### System Info :computer:
- OS: darwin
- Emacs: 25.2.1
- Spacemacs: 0.200.7
- Spacemacs branch: master (rev. c2774bc0)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
((auto-completion :variables haskell-completion-backend 'intero)
 emacs-lisp
 (haskell :variables haskell-enable-hindent-style "johan-tibell")
 helm osx ranger shell spell-checking syntax-checking vimscript)
```
- System configuration features: NOTIFY ACL LIBXML2 ZLIB TOOLKIT_SCROLL_BARS

#### Backtrace :paw_prints:
```
```

**Comment by scrotty** [comment]
I can't reproduce on my other computer with essentially the same setup.  It must be something on my end.

**Comment by troglotit** [comment]
Today I had a painful amount of freezes on html editing, with `:%s` command. This reproduction guide does reproduce freeze.
#### System Info :computer:
- OS: darwin
- Emacs: 25.1.1
- Spacemacs: 0.200.7
- Spacemacs branch: master (rev. c2774bc0)
- Graphic display: t
- Distribution: spacemacs
- Editing style: hybrid
- Completion: helm
- Layers:
```elisp
(csv auto-completion better-defaults colors dash docker git ibuffer org osx
     (ranger :variables ranger-show-preview nil)
     restclient search-engine
     (shell :variables shell-default-height 30 shell-default-position 'bottom shell-default-shell 'ansi-term)
     syntax-checking version-control clojure elm emacs-lisp go haskell javascript latex markdown ocaml python react
     (ruby :variables ruby-enable-enh-ruby-mode t ruby-version-manager 'rbenv)
     ruby-on-rails scala shell-scripts typescript vimscript yaml)
```
- System configuration features: JPEG RSVG NOTIFY ACL GNUTLS LIBXML2 ZLIB TOOLKIT_SCROLL_BARS NS MODULES

**Comment by scrotty** [comment]
Reopening since others seem to be encountering this issue too.

**Comment by bmag** [comment]
Could be Emacs choking on files with very long lines, or could be something else entirely... (maybe the search is implemented in efficiently?)

**Comment by scrotty** [comment]
I agree @bmag. I tend to edit large (many megabytes) files. I found (since opening this defect) many instances of search/replaces in spacemacs (possibly just emacs) that run extremely slowly or fail to finish at all. Doing the same operation on the same file in VIM, however, finishes nearly immediately.

I can't tell if its the "predictive" search/replace that's the problem (the feature that shows inline what will happen if you hit Enter - and a feature I flipping looove in spacemacs, btw!) or just the general search/replace.

**Comment by metalelf0** [comment]
I'm not sure this is totally related, but I'll post here then eventually open a new bug if anybody think it's needed. Sometimes when I'm doing a search spacemacs just hangs and brings CPU usage to 100%. It happens on both project-wide and file-wide searches. I'm using the emacs-plus package installed with homebrew as suggested in the installation guide, and I tried different emacs versions (25.1 - 25.2-rc1, 25-2rc2) without any difference. The issue happens on both develop and stable spacemacs branches.

I tried sending a `SIGUSR2` signal to spacemacs, and this is what I see in the debugger:

![](http://i.imgur.com/mn1mRKg.png)

Let me know if I can do anything else to help.

**Comment by Dessix** [comment]
I noticed that a huge amount of latency and soft-hangs came from redisplaying associated with highlighting, which may be related. In my case, I just did `/.` on a ~15k line file (C++ source) and waited, and every character was highlighted eventually, but it froze emacs once it started to get to a few hundred lines of highlighting every character. After about a minute, I was able to try highlighting something else to recover, but it seems that each character highlight triggers a redisplay of the entire line it's on. Could this be related?
